pkg/ipam/node.go

```diff
@@ -599,31 +596,84 @@ func (n *Node) maintainIPPool(ctx context.Context) error {
-		for _, ip := range a.release.IPsToRelease {
+	// Attempt to release if :
+	// * IP is still marked for release in the current iteration's a.release.IPsToRelease
+	// * Marked for release for more than excess-ip-release-delay seconds
+	// * IP is not in use by any pod according to pod informer cache
+	var ipsToRelease []string
+	if n.ipsMarkedForRelease != nil && len(n.ipsMarkedForRelease) > 0 {
+		if a.release == nil || len(a.release.IPsToRelease) == 0 {
+			n.ipsMarkedForRelease = make(map[string]time.Time)
+		}
+		for markedIp, ts := range n.ipsMarkedForRelease {
+			stillMarkedForRelease := false
+			for _, ip := range a.release.IPsToRelease {
+				if markedIp == ip {
+					stillMarkedForRelease = true
+					break
+				}
+			}
+			if !stillMarkedForRelease {
+				delete(n.ipsMarkedForRelease, markedIp)
+				continue
+			}
+			// Check if the IP release waiting period elapsed
+			if ts.Add(time.Duration(operatorOption.Config.ExcessIPReleaseDelay) * time.Second).After(time.Now()) {
+				continue
+			}
+			// Since, the agent's default CRD sync interval is 15 secs, there might be new allocations that aren't
+			// reflected in the ciliumnode object yet. Verify against pod cache to check if the IPs is in use.
 			if watchers.PodStore == nil {
 				scopedLog.Warnf("Pod store un-initialized")
 				continue
 			}
-			values, err := watchers.PodStore.(cache.Indexer).ByIndex(watchers.PodIpIndex, ip)
+			values, err := watchers.PodStore.(cache.Indexer).ByIndex(watchers.PodIpIndex, markedIp)
 			if err != nil {
 				scopedLog.WithError(err).Error("Unable to access pod store index")
 				continue
 			}
 			if len(values) != 0 {
-				scopedLog.Warnf("IP %s still in use for pod %s, not releasing", ip, values[0].(v1.Pod).Name)
-				continue
+				pod := values[0].(v1.Pod)
+				// Pods in "Failed" and "Succeeded" phases don't necessarily receive delete events.
+				// But their IPs can be released safely.
+				if k8sUtils.IsPodRunning(pod.Status) {
+					scopedLog.Warnf("IP %s still in use for pod %s, not releasing", markedIp, pod.Name)
+					continue
+				}
 			}
-			ipsToRelease = append(ipsToRelease, ip)
+			ipsToRelease = append(ipsToRelease, markedIp)
 		}
+	}
+
+	// Add entries from this iteration to ipsMarkedForRelease
+	releaseTS := time.Now()
+	if a.release != nil && a.release.IPsToRelease != nil {
+		for _, ip := range a.release.IPsToRelease {
+			if _, ok := n.ipsMarkedForRelease[ip]; !ok {
+				n.ipsMarkedForRelease[ip] = releaseTS
+			}
+		}
+	}
+
+	if len(ipsToRelease) > 0 {
 		a.release.IPsToRelease = ipsToRelease
+		scopedLog = scopedLog.WithFields(logrus.Fields{
+			"available":         n.stats.AvailableIPs,
+			"used":              n.stats.UsedIPs,
+			"excess":            n.stats.ExcessIPs,
+			"releasing":         a.release.IPsToRelease,
+			"selectedInterface": a.release.InterfaceID,
+			"selectedPoolID":    a.release.PoolID,
+		})
+		scopedLog.Info("Releasing excess IPs from node")
 		err := n.ops.ReleaseIPs(ctx, a.release)
```

Comment:
If this returns an error, we don't delete anything from `n.ipsMarkedForRelease`. So we'll try to re-release all these IPs next time round. If some IPs in the list were freed, will that call error out again because we try to free something we no longer own?

Is there a risk that a freed IP gets re-allocated somewhere else, but this node still has it in its "marked for release" list and subsequently tries to give it to a new pod on this node? (i.e. can an IP move back from "marked for release" to "let's use it" because in the meantime we got a new pod needing an IP?)